src/types/observableobject.ts

```diff
@@ -280,15 +282,24 @@ export class ObservableObjectAdministration
             const entry = this.pendingKeys_.get(key)
             if (entry) entry.set(true)
         }
-        this.keysAtom_.reportChanged()
+        this.reportKeysChanged()
     }
 
     getKeys_(): PropertyKey[] {
         this.keysAtom_.reportObserved()
+        if (!this.isStaledKeysValue_) {
+            return this.keysValue_
+        }
         // return Reflect.ownKeys(this.values) as any
         const res: PropertyKey[] = []
         for (const [key, value] of this.values_) if (value instanceof ObservableValue) res.push(key)
-        return res
+        this.isStaledKeysValue_ = false
+        return (this.keysValue_ = res)
```

Comment:
I am curious behind the rationale of this change. Did it solve a specific problem for you?

A risk of this change is that the keys are now shared amongst all readers, so if someone would `keys().sort().forEach()` for example, that would sort the keys for anybody requesting the keys. That could be protected against by using `Object.freeze(this.keysValue_)`.

Reply:
> I am curious behind the rationale of this change. Did it solve a specific problem for you?
I read the code, and think that it can be more efficient by using cache rather than retrieving `this.values_` every time.

> A risk of this change is that the keys are now shared amongst all readers, so if someone would keys().sort().forEach() for example, that would sort the keys for anybody requesting the keys. That could be protected against by using Object.freeze(this.keysValue_)
I prefer to use `slice()` to solve this issue.

Reply:
Yes, you're right. The `.slice()` is just a little bit faster by seeing [the bench](https://jsben.ch/ZDbO6).
I added `if(__DEV__) Object.freeze(this.keysValue_)` for avoiding consumer's modification.